### PR TITLE
Add pixelshuffle layers

### LINKED ISSUE #13548: [Feature Request / Proposal] Pixel shuffle layer
Upsampling based on pixel shuffling has been proposed in the paper [Real-Time Single Image and Video Super-Resolution Using an Efficient Sub-Pixel Convolutional Neural Network (2016)](https://arxiv.org/abs/1609.05158). The complete method combines a convolution layer, the pixel shuffle operation and a specific initialization to get rid of block artifacts. For the initialization details, see [Checkerboard artifact free sub-pixel convolution: A note on sub-pixel convolution, resize convolution and convolution resize (2017)](https://arxiv.org/abs/1707.02937).

Pixel shuffling in 2D means to reshape a tensor of shape `(N, f1*f2*C, H, W)` to `(N, C, f1*H, f2*W)`, thereby effectively upscaling the images by `(f1, f2)`.
In MXNet, pixel shuffling could be implemented like this in Python:

```python
def hybrid_forward(self, F, x):
        f1, f2 = self._factors
                                                      # (N, f1*f2*C, H, W)
        x = F.reshape(x, (0, -4, -1, f1 * f2, 0, 0))  # (N, C, f1*f2, H, W)
        x = F.reshape(x, (0, 0, -3, 0))               # (N, C, f1*f2*H, W)
        x = F.reshape(x, (0, 0, -4, -1, f2, 0))       # (N, C, f1*H, f2, W)
        x = F.reshape(x, (0, 0, 0, -3))               # (N, C, f1*H, f2*W)
        return x
```

Would this be an interesting addition?

### PR DESCRIPTION
**Pull request by kohr-h**
## Description ##
Pixel shuffle layers, as discussed in #13548 

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [ ] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

Closes #13548

### CODE REVIEW

File: python/mxnet/gluon/contrib/nn/basic_layers.py [hunk 1]
```diff
@@ -235,3 +235,135 @@ def _get_num_devices(self):
...
+            self._factors = (int(factor),) * 3
+        except TypeError:
+            self._factors = tuple(int(fac) for fac in factor)
+            assert len(self._factors) == 3, "wrong length {}".format(len(self._factors))
+
+    def hybrid_forward(self, F, x):
+        """Perform pixel-shuffling on the input."""
+        f1, f2, f3 = self._factors
+                                                              # (N, f1*f2*f3*C, D, H, W)
+        x = F.reshape(x, (0, -4, -1, f1 * f2 * f3, 0, 0, 0))  # (N, C, f1*f2*f3, D, H, W)
```
Comment (kohr-h) [inline python/mxnet/gluon/contrib/nn/basic_layers.py]: Does this always work? It generates a temporary 6D tensor, on CPU it's okay but I don't know about GPU.

**Comment by ThomasDelteil** [comment]
@kohr-h, thanks for the contribution! Can you add tests for your new layers? Also can you add a bit more documentation and reference the original paper, thanks!

File: python/mxnet/gluon/contrib/nn/basic_layers.py (hunk 1 above)
Comment (ThomasDelteil) [reply python/mxnet/gluon/contrib/nn/basic_layers.py]: Adding tests will let you test that automatically on all our supported platform when your PR is being tested in the CI. GPU / CPU / Windows / Linux / arm / etc.

**Comment by Roshrini** [comment]
@mxnet-label-bot Update [Operator, pr-awaiting-response]

**Comment by kohr-h** [comment]
Axis swapping issue is fixed now. The 2D version now does the same as the PyTorch implementation. I'll add some more documentation.

**Comment by kohr-h** [comment]
Just updated the superresolution example with the new layer, seems to work fine. The old download link wasn't working any more, so I switched to a GitHub mirror of the newer version of the dataset.

**Comment by kohr-h** [comment]
Ready for review now.

**Comment by Roshrini** [comment]
@apeforest @zhreshold @eric-haibin-lin Can you please help in reviewing this PR? Thanks
@mxnet-label-bot Update [pr-awaiting-review, Operator]

**Comment by sandeep-krishnamurthy** [comment]
@ThomasDelteil - Can you please help review this PR?

**Comment by vandanavk** [comment]
LGTM.

@zhreshold @Mut1nyJD for review.

@kohr-h if merging this PR would fix https://github.com/apache/incubator-mxnet/issues/13548, please add "Fixes https://github.com/apache/incubator-mxnet/issues/13548" in the PR description. this will automatically close the issue when the PR is merged.

**Comment by vandanavk** [comment]
@mxnet-label-bot add [Gluon]

**Comment by kohr-h** [comment]
@vandanavk I added the corresponding text.

**Comment by ankkhedia** [comment]
@ThomasDelteil Could you please help review/merge this PR?

**Comment by kohr-h** [comment]
:+1:
